Source/Concurrency/CivlTypeChecker.cs

```diff
@@ -1402,17 +1436,32 @@ public override Expr VisitIdentifierExpr(IdentifierExpr node)
               node.Decl.Name);
           }
         }
-
         return base.VisitIdentifierExpr(node);
       }
-
+      
       public override Cmd VisitCallCmd(CallCmd node)
       {
-        civlTypeChecker.Error(node, "Call command not allowed inside an atomic action");
+        if (!actionProcToLayerRange.ContainsKey(node.Proc))
+        {
+          civlTypeChecker.Error(node, "An atomic action can only call other atomic actions");
+        }
+        else if (!actionProcToLayerRange[proc].Subset(actionProcToLayerRange[node.Proc]))
+        {
+          civlTypeChecker.Error(node, "Caller layer range must be subset of callee layer range");
+        }
+        else
+        {
+          callGraph.AddEdge(proc, node.Proc);
+        }
         return base.VisitCallCmd(node);
       }
```

Comment:
There is a soundness issue here where an atomic action can call an introduction action, thereby getting access to a global variable a layer to early. Imagine atomic action `A`, introduction action `I`, and global variable `g`, all with singleton layer range `[1]`. Then `I` is allowed to access `g`, but `A` is not. However, if `A` calls `I` then the access to `g` will get inlined into `A`.

Reply:
Added another case in the type checking above.  Also added a test case.